src/map/mob.cpp

```diff
@@ -507,42 +495,42 @@ struct mob_data* mob_spawn_dataset(struct spawn_data *data)
  *------------------------------------------*/
 int mob_get_random_id(int type, enum e_random_monster_flags flag, int lv)
 {
-	struct s_randomsummon_group *msummon = (struct s_randomsummon_group *)idb_get(mob_summon_db, type);
+	std::shared_ptr<s_randomsummon_group> summon = mob_summon_db.find(type);
 
 	if (type == MOBG_Bloody_Dead_Branch && flag&RMF_MOB_NOT_BOSS)
 		flag = static_cast<e_random_monster_flags>(flag&~RMF_MOB_NOT_BOSS);
 	
-	if (!msummon) {
+	if (!summon) {
 		ShowError("mob_get_random_id: Invalid type (%d) of random monster.\n", type);
 		return 0;
 	}
-	if (!msummon->count) {
+	if (summon->list.empty()) {
 		ShowError("mob_get_random_id: Random monster type %d is not defined.\n", type);
 		return 0;
 	}
 
 	std::shared_ptr<s_mob_db> mob;
-	int i = 0, mob_id = 0, rand = 0;
-	struct s_randomsummon_entry *entry = nullptr;
+	uint32 i = 0;
+	uint16 mob_id = 0;
+	std::shared_ptr<s_randomsummon_entry> entry = nullptr;
 
 	do {
-		rand = rnd()%msummon->count;
-		entry = &msummon->list[rand];
+		entry = util::umap_random(summon->list);
 		mob_id = entry->mob_id;
 		mob = mob_db.find(mob_id);
-	} while ((rand == 0 || // Skip default first
+	} while ((
 		mob == nullptr ||
 		mob_is_clone(mob_id) ||
 		(flag&RMF_DB_RATE && (entry->rate < 1000000 && entry->rate <= rnd() % 1000000)) ||
 		(flag&RMF_CHECK_MOB_LV && lv < mob->lv) ||
 		(flag&RMF_MOB_NOT_BOSS && status_has_mode(&mob->status,MD_STATUSIMMUNE) ) ||
 		(flag&RMF_MOB_NOT_SPAWN && !mob_has_spawn(mob_id)) ||
 		(flag&RMF_MOB_NOT_PLANT && status_has_mode(&mob->status,MD_IGNOREMELEE|MD_IGNOREMAGIC|MD_IGNORERANGED|MD_IGNOREMISC) )
-	) && (i++) < MAX_MOB_DB && msummon->count > 1);
+	) && (i++) < MAX_MOB_DB);
```

Comment:
I do not know why someone used the MAX_MOB_DB constant here, but we should probably replace this.
I think in the other cases I did something like .size() * 3
I think the best would be to replace this whole chaos with a normal for loop.

Most likely like this (I did not compile this to try):
```
for( size_t i = 0, max = summon->list.size() * 3; i < max; i++ ){
	std::shared_ptr<s_randomsummon_entry> entry = util::umap_random( summon->list );
	std::shared_ptr<s_mob_db> mob = mob_db.find( entry->mob_id );

	if( 	mob == nullptr ||
		mob_is_clone( entry->mob_id ) ||
		(flag&RMF_DB_RATE && (entry->rate < 1000000 && entry->rate <= rnd() % 1000000)) ||
		(flag&RMF_CHECK_MOB_LV && lv < mob->lv) ||
		(flag&RMF_MOB_NOT_BOSS && status_has_mode(&mob->status,MD_STATUSIMMUNE) ) ||
		(flag&RMF_MOB_NOT_SPAWN && !mob_has_spawn( entry->mob_id )) ||
		(flag&RMF_MOB_NOT_PLANT && status_has_mode(&mob->status,MD_IGNOREMELEE|MD_IGNOREMAGIC|MD_IGNORERANGED|MD_IGNOREMISC) )
	){
		continue;
	}

	return entry->mob_id;
}

return summon->default_mob_id;
```

Reply:
Should we check if
```
std::shared_ptr<s_mob_db> mob = mob_db.find( summon->default_mob_id );
```
can be nullptr as well?

Reply:
You can add it for safety if you want. :-)